dg.go

```diff
@@ -1,315 +1,480 @@
-            id,
-        }
-    }
-    return n, nil
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	n, ok := d.nodes[id]
+	if !ok {
+		return nil, &ErrNodeNotFound{
+			id,
+		}
+	}
+	return n, nil
 }
 
 func (d *directedGraph[NodeType]) ListNodes() map[string]Node[NodeType] {
-    d.lock.Lock()
-    defer d.lock.Unlock()
-
-    result := map[string]Node[NodeType]{}
-    for nodeID, n := range d.nodes {
-        result[nodeID] = n
-    }
-    return result
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	result := map[string]Node[NodeType]{}
+	for nodeID, n := range d.nodes {
+		result[nodeID] = n
+	}
+	return result
 }
 
 func (d *directedGraph[NodeType]) ListNodesWithoutInboundConnections() map[string]Node[NodeType] {
-    d.lock.Lock()
-    defer d.lock.Unlock()
-
-    result := map[string]Node[NodeType]{}
-    for nodeID, n := range d.nodes {
-        if len(d.connectionsToNode[nodeID]) == 0 {
-            result[nodeID] = n
-        }
-    }
-    return result
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	result := map[string]Node[NodeType]{}
+	for nodeID, n := range d.nodes {
+		connections := d.connectionsToNode[nodeID]
+		if len(connections) == 0 {
+			result[nodeID] = n
+		}
+	}
+	return result
+}
+
+// Validates the specified node IDs and confirms that a connection between them
+// would be valid, then sets the `to` and `from` connections and adds the
+// dependency to the `to` node.
+func (d *directedGraph[NodeType]) connectNodes(fromID, toID string, dependencyType DependencyType) error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	// Make sure both nodes exist and are not deleted.
+	fromNode, ok := d.nodes[fromID]
+	if !ok {
+		return &ErrNodeNotFound{fromID}
+	} else if fromNode.deleted {
+		return &ErrNodeDeleted{fromID}
+	}
+	toNode, ok := d.nodes[toID]
+	if !ok {
+		return &ErrNodeNotFound{toID}
+	} else if toNode.deleted {
+		return &ErrNodeDeleted{toID}
+	}
+	// Check that it's a non-self and non-duplicate connection.
+	if fromID == toID {
+		return &ErrCannotConnectToSelf{fromID}
+	}
+	if _, ok := d.connectionsFromNode[fromID][toID]; ok {
+		return &ErrConnectionAlreadyExists{fromID, toID}
+	}
+	// Update the mappings.
+	d.connectionsFromNode[fromID][toID] = struct{}{}
+	d.connectionsToNode[toID][fromID] = struct{}{}
+	// Update the dependencies
+	toNode.outstandingDependencies[fromID] = dependencyType
+	return nil
+}
+
+func (d *directedGraph[NodeType]) PushStartingNodes() error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	for nodeID, n := range d.nodes {
+		if len(n.outstandingDependencies) == 0 {
+			d.readyForProcessing[nodeID] = n
+		}
+	}
+	return nil
+}
+
+func (d *directedGraph[NodeType]) PopReadyNodes() []*node[NodeType] {
+	d.lock.Lock()
+	// Transfer the map to a local variable to minimize time locked, and reset the graph's value.
+	readyMap := d.readyForProcessing
+	d.readyForProcessing = make(map[string]*node[NodeType])
+	d.lock.Unlock()
+	// Copy the map nodes to a slice to comply with the interface.
```

Comment:
When I'm reading this code, telling me that the interface calls for a slice doesn't provide any value: that's obvious. What *would* provide value is explaining *why* the interface calls for a slice. (The interface already covers that, more or less, and anything here probably doesn't add anything worthwhile.)